### PR TITLE
Feature/model lighting

### PR DESCRIPTION
**Pull request by emiax**
- Remove hard coded sun light source for RenderableModel
- Introduce LightSource interface
- Implement CameraLightSource, to always follow camera
- Implement SceneGraphLightSource to follow scene graph node
- Change model shader to accept up to 8 light sources
- Add properties to RenderableModel for diffuse, ambient and specular intensities
- Fix New Horizons label texture, so that the colors align with the text

All light sources are directional at this point. Keeping it simple. This may need to be redone if we go more into the deferred shading direction, but this should be useful until we get there.

### CODE REVIEW

File: modules/base/lightsource/cameralightsource.cpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+ ****************************************************************************************/
+
+#include <modules/base/lightsource/cameralightsource.h>
+
+#include <openspace/documentation/documentation.h>
+#include <openspace/documentation/verifier.h>
+#include <openspace/util/updatestructures.h>
+
+namespace {
+    constexpr const openspace::properties::Property::PropertyInfo IntensityInfo = {
```
Comment (alexanderbock) [inline modules/base/lightsource/cameralightsource.cpp]: If I remember correctly, just `constexpr` will do the same a `constexpr const`

File: modules/base/lightsource/cameralightsource.cpp [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        );
+    }
+}
+
+float CameraLightSource::intensity() const {
+    return _intensity;
+}
+
+glm::vec3 CameraLightSource::directionViewSpace(const RenderData& renderData) const {
+    return glm::vec3(0.0, 0.0, 1.0);
```
Comment (alexanderbock) [inline modules/base/lightsource/cameralightsource.cpp]: `0.f, 0.f, 1.f`

File: modules/base/lightsource/cameralightsource.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+public:
+    CameraLightSource();
+    CameraLightSource(const ghoul::Dictionary& dictionary);
+
+    static documentation::Documentation Documentation();
+
+    glm::vec3 directionViewSpace(const RenderData& renderData) const override;
+
+    float intensity() const override;
+private:
```
Comment (alexanderbock) [inline modules/base/lightsource/cameralightsource.h]: Style:  Extra empty line before `private:`

File: modules/base/lightsource/cameralightsource.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ *                                                                                       *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,   *
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A         *
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT    *
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF  *
+ * CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE  *
+ * OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                                         *
+ ****************************************************************************************/
+
+#ifndef __OPENSPACE_MODULE_BASE___CAMREALIGHTSOURCE___H__
```
Comment (alexanderbock) [inline modules/base/lightsource/cameralightsource.h]: CAMREA

File: modules/base/lightsource/scenegraphlightsource.cpp [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+#include <openspace/documentation/documentation.h>
+#include <openspace/documentation/verifier.h>
+#include <openspace/engine/openspaceengine.h>
+#include <openspace/rendering/renderengine.h>
+#include <openspace/scene/scene.h>
+#include <openspace/util/updatestructures.h>
+
+namespace {
+    constexpr const openspace::properties::Property::PropertyInfo IntensityInfo = {
```
Comment (alexanderbock) [inline modules/base/lightsource/scenegraphlightsource.cpp]: Same as above

File: modules/base/lightsource/scenegraphlightsource.cpp [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    return _sceneGraphNode != nullptr;
+}
+
+float SceneGraphLightSource::intensity() const {
+    return _intensity;
+}
+
+glm::vec3 SceneGraphLightSource::directionViewSpace(const RenderData& renderData) const {
+    if (!_sceneGraphNode) {
+        return glm::vec3(0.0);
```
Comment (alexanderbock) [inline modules/base/lightsource/scenegraphlightsource.cpp]: `0.f`

File: modules/base/lightsource/scenegraphlightsource.cpp [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    glm::dvec3 lightPosition = glm::vec4(
```
Comment (alexanderbock) [inline modules/base/lightsource/scenegraphlightsource.cpp]: conversion to float glm::vec4 on purpose?

File: modules/base/lightsource/scenegraphlightsource.cpp [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        _sceneGraphNode->modelTransform() * glm::dvec4(0.0, 0.0, 0.0, 1.0)
+    );
+
+    glm::dvec3 renderNodePosition = renderData.modelTransform.translation;
```
Comment (alexanderbock) [inline modules/base/lightsource/scenegraphlightsource.cpp]: `const`

File: modules/base/lightsource/scenegraphlightsource.cpp [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+
+    glm::dvec3 lightDirectionViewSpace = renderData.camera.viewRotationMatrix() *
```
Comment (alexanderbock) [inline modules/base/lightsource/scenegraphlightsource.cpp]: `const`

File: modules/base/lightsource/scenegraphlightsource.h [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    glm::vec3 directionViewSpace(const RenderData& renderData) const override;
+
+    float intensity() const override;
+
+    bool initialize() override;
+private:
+    properties::FloatProperty _intensity;
+    properties::StringProperty _sceneGraphNodeReference;
+
+    SceneGraphNode* _sceneGraphNode;
```
Comment (alexanderbock) [inline modules/base/lightsource/scenegraphlightsource.h]: ` = nullptr`

File: modules/base/rendering/renderablemodel.cpp [hunk 11]
```diff
@@ -134,22 +194,58 @@ RenderableModel::RenderableModel(const ghoul::Dictionary& dictionary)
...
 }
 
+RenderableModel::~RenderableModel() {}
+
 bool RenderableModel::isReady() const {
     return _programObject && _texture;
 }
 
+void RenderableModel::initialize() {
+    for (auto& ls : _lightSources) {
```
Comment (alexanderbock) [inline modules/base/rendering/renderablemodel.cpp]: No need for `auto`

File: modules/base/rendering/renderablemodel.cpp [hunk 12]
```diff
@@ -213,15 +295,31 @@ void RenderableModel::render(const RenderData& data, RendererTasks&) {
...
 
-    const glm::vec3 directionToSun = glm::normalize(
-        _sunPos - data.modelTransform.translation
-    );
-    const glm::vec3 directionToSunViewSpace =
-        glm::normalize(glm::mat3(data.camera.combinedViewMatrix()) * directionToSun);
+    int nLightSources = 0;
+    _lightIntensitiesBuffer.resize(_lightSources.size());
+    _lightDirectionsViewSpaceBuffer.resize(_lightSources.size());
+    for (const auto& lightSource : _lightSources) {
```
Comment (alexanderbock) [inline modules/base/rendering/renderablemodel.cpp]: No need for `auto`

File: src/scene/lightsource.cpp [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+#include <ghoul/logging/logmanager.h>
+#include <ghoul/misc/dictionary.h>
+#include <ghoul/misc/templatefactory.h>
+
+namespace {
+    constexpr const char* KeyType = "Type";
+
+    constexpr const char* KeyIdentifier = "Identifier";
+
+    constexpr const openspace::properties::Property::PropertyInfo EnabledInfo = {
```
Comment (alexanderbock) [inline src/scene/lightsource.cpp]: See above

File: src/scene/lightsource.cpp [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+ * CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE  *
+ * OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                                         *
+ ****************************************************************************************/
+
+#include <openspace/scene/lightsource.h>
+
+#include <openspace/documentation/documentation.h>
+#include <openspace/documentation/verifier.h>
+#include <openspace/util/factorymanager.h>
+
```
Comment (alexanderbock) [inline src/scene/lightsource.cpp]: You can condense all of the includes;  only the space after the `#include <openspace/scene/lightsource.h> is necessary

File: src/scene/lightsource.cpp [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+        "Enabled",
+        "Enabled",
+        "Whether the light source is enabled or not"
+    };
+} // namespace
+
+namespace openspace {
+
+bool LightSource::isEnabled() const
+{
```
Comment (alexanderbock) [inline src/scene/lightsource.cpp]: `{` on the same line as function

File: src/scene/lightsource.cpp [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+documentation::Documentation LightSource::Documentation() {
+    using namespace openspace::documentation;
+
+    return {
+        "Light Source",
+        "core_light_source",
+        {
+
```
Comment (alexanderbock) [inline src/scene/lightsource.cpp]: Extra empty line

File: src/scene/lightsource.cpp [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+                EnabledInfo.identifier,
+                new BoolVerifier,
+                Optional::Yes,
+                EnabledInfo.description
+            }
+        }
+    };
+}
+
+std::unique_ptr<LightSource> LightSource::createFromDictionary(const ghoul::Dictionary& dictionary) {
```
Comment (alexanderbock) [inline src/scene/lightsource.cpp]: This line seems too long

File: modules/base/lightsource/cameralightsource.h (hunk 4 above)
Comment (emiax) [reply modules/base/lightsource/cameralightsource.h]: Well spotted!

**Pull request merged**